sdks/python/apache_beam/metrics/metricbase.py

```diff
@@ -78,7 +80,7 @@ def __hash__(self):
 
 class Metric(object):
   """Base interface of a metric object."""
-  pass
+  metric_name = None  # type: MetricName
```

Comment:
there seems to be an assumption that this attribute exists.  should we give this class an `__init__` that takes a `metric_name` argument?

Reply:
I think it exists due to some double inheritance. @ajamato what is the best solution here?

Reply:
Its not due to multiple inheritance, its just self.metric_name is simply not set in the base Metric class, but only in the DelegatingCounter, DelegatingDistribution and DelegatingGauge class. (Which have an __init__ which takes a metric_name and assigns it to self). The rest of the code in the repo is assuming self.metric_name is set, as they only deal with the subclasses.

I recommend adding __init__ with a metric_name parameter to Metric and its subclasses (Counter, Distribution and Gauge). This already exists on DelegatingX classes, but could be done in the base Metric __init__ instead.

Also,
If it were up to me we would remove Counter, Distribution and Gauge classes and just have the DelegatingCounter DelegatingDistribution and DelegatingGauge classes.

Though, I am unsure if some external implementation to the beam repo uses those classes (Counter, Distribution and Gauge).

Reply:
>I recommend adding init with a metric_name parameter to Metric and its subclasses (Counter, Distribution and Gauge). This already exists on DelegatingX classes, but could be done in the base Metric init instead.

Adding `metric_name` to `Metric.__init__` definitely seems like the best solution to me, so I did that. 

> If it were up to me we would remove Counter, Distribution and Gauge classes and just have the DelegatingCounter DelegatingDistribution and DelegatingGauge classes.

Leaving this for a future task.

Reply:
Sounds good. I wouldn't want to scope creep that into your PR. Definitely for another day :).